stablehlo/reference/Api.cpp

```diff
@@ -170,19 +170,82 @@ LogicalResult removeDynamism(ModuleOp module, func::FuncOp func,
   return success();
 }
 
+bool isQuantizedTypes(TypeRange types) {
+  return llvm::any_of(types, [](Type type) {
+    return isa<quant::QuantizedType>(getElementTypeOrSelf(type));
+  });
+}
+
+// Recursively checks if an operation or any of its nested operations use
+// quantized types.
+//
+// Args:
+//   op: The operation to check for quantized type usage.
+//
+// Returns:
+//   True if the operation or any nested operation uses quantized types,
+//   false otherwise.
+bool operationUsesQuantType(Operation &op) {
+  for (Region &region : op.getRegions()) {
+    for (Block &block : region) {
+      for (Operation &op : block) {
+        TypeRange operandTypes = op.getOperandTypes();
+        TypeRange resultTypes = op.getResultTypes();
+        if (isQuantizedTypes(operandTypes) || isQuantizedTypes(resultTypes) ||
+            operationUsesQuantType(op)) {
+          return true;
+        }
+      }
+    }
+  }
+  return false;
+}
+
+// Lowers quantization-related operations and types within a function to
+// primitive math operations.
+//
+// This function checks if a function uses quantized types in its inputs,
+// outputs, or internal operations. If so, it creates and runs a StableHLO
+// quantization lowering pipeline to transform those quantized constructs into
+// primitive math operations. If the lowering process fails, an error is
+// emitted.
+//
+// Args:
+//   module: The module containing the function `func`.
+//   func The function to lower quantized types/operations in.
+//
+// Returns:
+//   A `LogicalResult` indicating success or failure of the lowering process.
+LogicalResult lowerQuantization(ModuleOp module, func::FuncOp func) {
+  if (!(isQuantizedTypes(func.getFunctionType().getInputs()) ||
+        operationUsesQuantType(*func.getOperation()) ||
+        isQuantizedTypes(func.getFunctionType().getResults()))) {
```

Comment:
I think this might be redundant? I would expect the walking in `operationUsesQuantType` to hit the function inputs as the arguments to the function's body (block), and the return types as the results of the block?

Reply:
PTAL at the revised code. 
Do you mean checking the block arguments and return types for all the operations?
Given that in stablehlo most of the operations (except for if/case/while) are single block operations, I skipped that using  a special check for function op